### PR TITLE
Electrical Meter view

### PR DESCRIPTION
**Pull request by aleemont1**
TODO:
- [x] Implement Add/Remove Outlets Buttons
- [x] Update UI in real time
- [x] Automatic scaling

### CODE REVIEW

**Comment by aleemont1** [comment]
@smagnaniuni @michelececcacci I'd like to have your feedback on the developing of these classes.
It's in a "usable" state. I'll need to study a better way of updating it during the simulation, and some way to display labels for Outlets'  names, and eventually add a column wich indicates the plugged device, but it's secondary.

I need to know if this way of developing the view is ok, if I need to make adjustments and/or if I'm doing errors somewhere which I didn't notice.

**Comment by michelececcacci** [comment]
I can't actually see the view

File: src/main/java/homer/view/javafx/ElectricalMeterView.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        outletTable.setEditable(false);
+        outletTable.setItems(outlets);
+
+        outletStatusColumn = new TableColumn<>("Status");
+        outletStatusColumn.setCellFactory(param -> new TableCell<List<Outlet>, Double>() {
+            @Override
+            protected void updateItem(final Double item, final boolean empty) {
+                super.updateItem(item, empty);
+                if (empty) {
+                    setText(null);
```
[the remark this answers is not in the document]
Comment (aleemont1) [reply src/main/java/homer/view/javafx/ElectricalMeterView.java]: This class was deleted in 068b52e35f21caa1f69a0ceb9ea0bf25a0220fba

File: src/main/java/homer/view/javafx/ElectricalMeterView.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                } else {
+                    TableRow<List<Outlet>> row = getTableRow();
+                    if (row != null) {
```
[the remark this answers is not in the document]
Comment (aleemont1) [reply src/main/java/homer/view/javafx/ElectricalMeterView.java]: As above.

File: src/main/java/homer/view/LaunchMeterView.java [hunk 3]
```diff
@@ -69,23 +70,26 @@ public void start(final Stage stage) throws Exception {
...
+                light.updateTick(Duration.ofMillis(millis));
+                outlet.getState().addValue(light.getInstantConsumption());
+                outlet.updateTick(Duration.ofHours(hours));
+            }
+            meter.updateTick(Duration.ofHours(hours));
+            view.setLabels();
+        }));
+        timeline.setCycleCount(Timeline.INDEFINITE);
+        timeline.play();
+
```
Comment (aleemont1) [inline src/main/java/homer/view/LaunchMeterView.java:92]: Updates the view every 300ms (about 3 fps) indefinitely.
In particular the view is updated by calling `view.setLabels()`.
This is just a showcase for test purposes, the real implementation of this code is going to be in the main loop.

Looks like it's working really well, better then I expected actually.

Though I don't really like how `Outlet`'s state must be updated before calling `updateTIck()`, maybe I should instantiate Outlets with a default state value, which should be a little over 0.0 (e.g. 0.0001 should be enough), but I'd rather not hardcode it. I'll go through it and hope there's no need to heavily refactor `Outlets`/`PoweredDevice`/`PoweredDeviceInfo`. 

Using a `Timeline` object might be useful for updating the view indipendently from the rest of the loop, it updates the view at a fixed frequency, and if run on another thread shouldn't be slowed down by the code (Shouldn't become unresponsive during long calcs, and would be updated piece by piece).

@smagnaniuni @michelececcacci what do you think?

Comment (smagnaniuni) [reply src/main/java/homer/view/LaunchMeterView.java:92]: Should be good, was thinking of using a timeline for the graphs too but have no idea how to use it at the moment.
Have to review later

**Comment by michelececcacci** [comment]
now gh cli seems to work, i'll check this out tonight or tomorrow.

File: src/main/java/homer/view/javafx/sensorsview/SensorDashboardViewManager.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    // Reference of PM25Label for the FXML loader.
+    private Label pm25Label;
+
+    @FXML
+    // Reference of CO2Label for the FXML loader.
+    private Label co2Label;
+
+    @FXML
+    // Reference of ToxicGasLabel for the FXML loader.
+    private Label toxicGasLabel;
```
Comment (michelececcacci) [inline src/main/java/homer/view/javafx/sensorsview/SensorDashboardViewManager.java:44]: aren't we supposed to store air quality info in the main view? Just asking, but it's perfectly fine if you need to. Just check out the changes introduced by #200 if you actually need the values.

**Review by michelececcacci — COMMENTED** [review]
can't checkout on the pr itself, but i'll try my best to keep up to date.

File: src/main/java/homer/view/javafx/sensorsview/SensorDashboardViewManager.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    /**
+     * Sets {@code environment} to a new
+     * {@link homer.model.environment.HomeEnvironment}.
+     * 
+     * @param environment the new {@code HomeEnvironment}.
+     */
+    public void setEnvironment(final HomeEnvironment environment) {
+        Objects.requireNonNull(environment);
```
Comment (michelececcacci) [inline src/main/java/homer/view/javafx/sensorsview/SensorDashboardViewManager.java:92]: I think you can do this in just one line. This should work
```java
        
        this.environment = new HomeEnvironment(Objects.requireNonNull(environment));
```

File: src/main/java/homer/view/javafx/sensorsview/SensorDashboardViewManager.java (hunk 4 above)
Comment (aleemont1) [reply src/main/java/homer/view/javafx/sensorsview/SensorDashboardViewManager.java:44]: As the comment says, it's just a reference for the FXML loader.

Comment (michelececcacci) [reply src/main/java/homer/view/javafx/sensorsview/SensorDashboardViewManager.java:44]: Oh, ok. I never really used FXML, so i thought this were like opt-in.

**Comment by aleemont1** [comment]
I know the update of the labels is way too fast, but it's working and functional, I'll fix it later if we have enough time. Will start working on the other 2 views now.

**Pull request merged**